storage/reads/array_cursor.gen.go.tmpl

```diff
@@ -242,16 +256,84 @@ func (c *{{.name}}LimitArrayCursor) Next() {{$arrayType}} {
 	return c.res
 }
 
-type {{.name}}WindowLimitArrayCursor struct {
+type {{.name}}WindowLastArrayCursor struct {
+	cursors.{{.Name}}ArrayCursor
+	every int64
+	final int64
+	res {{$arrayType}}
+	tmp {{$arrayType}}
+}
+
+func new{{.Name}}WindowLastArrayCursor(cur cursors.{{.Name}}ArrayCursor, every int64) *{{.name}}WindowLastArrayCursor {
+	return &{{.name}}WindowLastArrayCursor{
+		{{.Name}}ArrayCursor: cur,
+		every: every,
+		final: math.MinInt64,
+		res:   cursors.New{{.Name}}ArrayLen(MaxPointsPerBlock),
+		tmp:   &cursors.{{.Name}}Array{},
+	}
+}
+
+func (c *{{.name}}WindowLastArrayCursor) Stats() cursors.CursorStats {
+	return c.{{.Name}}ArrayCursor.Stats()
+}
+
+func (c *{{.name}}WindowLastArrayCursor) Next() *cursors.{{.Name}}Array {
+	every := c.every
+	if every == 0 {
+		every = math.MaxInt64
+	}
+
+	cur := -1
+
+NEXT:
+	var a *cursors.{{.Name}}Array
+
+	if c.tmp.Len() > 0 {
+		a = c.tmp
+	} else {
+		a = c.{{.Name}}ArrayCursor.Next()
+	}
+
+	if a.Len() == 0 {
+		c.res.Timestamps = c.res.Timestamps[:cur+1]
+		c.res.Values = c.res.Values[:cur+1]
+		return c.res
+	}
+
+	for i, t := range a.Timestamps {
+		if t >= c.final {
+			cur++
+		}
+
+		if cur == MaxPointsPerBlock {
+			c.tmp.Timestamps = a.Timestamps[i+1:]
+			c.tmp.Values = a.Values[i+1:]
+			return c.res
+		}
+
+		c.res.Timestamps[cur] = t
+		c.res.Values[cur] = a.Values[i]
+
+		c.final = t - t % every + every
+	}
+
+	c.tmp.Timestamps = nil
+	c.tmp.Values = nil
+
+	goto NEXT
+}
+
+type {{.name}}WindowFirstArrayCursor struct {
 	cursors.{{.Name}}ArrayCursor
 	every int64
 	final int64
 	res {{$arrayType}}
 	tmp {{$arrayType}}
 }
 
-func new{{.Name}}WindowLimitArrayCursor(cur cursors.{{.Name}}ArrayCursor, every int64) *{{.name}}WindowLimitArrayCursor {
-	return &{{.name}}WindowLimitArrayCursor{
+func new{{.Name}}WindowFirstArrayCursor(cur cursors.{{.Name}}ArrayCursor, every int64) *{{.name}}WindowFirstArrayCursor {
+	return &{{.name}}WindowFirstArrayCursor{
```

Comment:
What did limit do?

Reply:
Previously there was only a single "limit" array cursor for handling both last and first operations. But windowing needs different cursor implementations for first and last.

Reply:
Okay, yeah that makes more sense now.